### PR TITLE
Add gas turbine model, part of NGCC

### PR DESCRIPTION
**Pull request by eslickj**
This adds the gas turbine model and tests.  It may make more sense as an example, but this seems to be the standard way to handle flowsheets for now.  This branches from PR #301, so that should be merged first.  It should be safe to review this though.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by eslickj** [comment]
@andrewlee94, I'm not sure how, but I think somehow the test I wrote for the gas turbine is getting tangled up in other tests.  The model runs fine for me and the test runs fine for me by itself, but on here it failing with 

```
 /home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/tests/test_gas_turbine.py:70: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/gas_turbine.py:397: in main
    tags, tag_format = tag_model(m)
/home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/gas_turbine.py:61: in tag_model
    new_tag(f"{i}_Fmol", expr=s.flow_mol/1000, format="{:.3f} kmol/s")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <idaes.generic_models.properties.examples.saponification_thermo.SaponificationStateBlockData object at 0x7faaf1d9a510>
attr = 'flow_mol'
``` 

I'm not using the saponification package.  I'm not really sure how this is happening.

**Comment by andrewlee94** [comment]
> @andrewlee94, I'm not sure how, but I think somehow the test I wrote for the gas turbine is getting tangled up in other tests. The model runs fine for me and the test runs fine for me by itself, but on here it failing with
> 
> ```
>  /home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/tests/test_gas_turbine.py:70: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> /home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/gas_turbine.py:397: in main
>     tags, tag_format = tag_model(m)
> /home/runner/work/idaes-pse/idaes-pse/idaes/power_generation/flowsheets/gas_turbine/gas_turbine.py:61: in tag_model
>     new_tag(f"{i}_Fmol", expr=s.flow_mol/1000, format="{:.3f} kmol/s")
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> self = <idaes.generic_models.properties.examples.saponification_thermo.SaponificationStateBlockData object at 0x7faaf1d9a510>
> attr = 'flow_mol'
> ```
> 
> I'm not using the saponification package. I'm not really sure how this is happening.

I have no idea how this is happening either. I just ran the tests locally myself and they pass. A quick look over the code also shows no obvious way this could happen.

**Comment by eslickj** [comment]
@andrewlee94 It doesn't make a lot of sense, but my feeling is that we are getting our wires crossed with a global variable somehow.  I can't imagine how it's happening though.  I'm going to try to track this down.

**Comment by eslickj** [comment]
@andrewlee94 for the test, I found that there is an extra stream called "model.stream".  After searching, the only place that could come from is the tables tests.  I also added an assert on the model name, so I know I have the gas turbine model with elements of the tables test fixture model mixed into it.  Still no idea how that could happen, but at least it narrows it down a lot.

**Comment by eslickj** [comment]
@andrewlee94, yeah I looked more, and that must be the case.  The stream table from the table test is somehow held over, and added to by the gas turbine.  I sure don't see how, but at least we have it pinned down.

**Comment by eslickj** [comment]
@andrewlee94 looks like I fixed it.  I don't quite understand what's going on, but the issue was in ```stream_states_dict()```.  There is a subfunction, and somehow it seems to create it's own persistent copy of the stream_states dict.  I guess the subfunction was making a copy of the stream_states dict on the first go then using it's own copy after that.  I clearly don't understand persistence in Python, but now I worried we could have the same issue elsewhere.

**Comment by andrewlee94** [comment]
@eslickj Was this a sub-function in your code, or something in the core IDAES code? I.e. do we need to dig in deeper and fix the underlying issue.

**Comment by eslickj** [comment]
@andrewlee94 it was in the IDAES core, but that's where I fixed it.  My only worry is that we could have made the same mistake in other places, and it could be pretty tough to spot and cause very mysterious problems.  I wounder if there is a tool to find inner-functions.  Then we could more easily check them all for the same issue.

**Comment by eslickj** [comment]
@andrewlee94, okay I found the real problem.  Basically this:

```python
def f(k, x={}):
    x[k] = "Hi"
    return x

> print(f(1))
{1: "Hi"}
> print(f(2))
{1: "Hi", 2: "Hi"}
```

If you keep the reference to x around the next time you call the function, the x arg is not an empty dict, but the same dict that was returned by the function.  Kind of interesting.  I guess this is not usually a problem because if the function doesn't return x, the x reference would disappear at the end of the function.

I was stating to question reality.

**Comment by andrewlee94** [comment]
> @andrewlee94, okay I found the real problem. Basically this:
> 
> ```python
> def f(k, x={}):
>     x[k] = "Hi"
>     return x
> 
> > print(f(1))
> {1: "Hi"}
> > print(f(2))
> {1: "Hi", 2: "Hi"}
> ```
> 
> If you keep the reference to x around the next time you call the function, the x arg is not an empty dict, but the same dict that was returned by the function. Kind of interesting. I guess this is not usually a problem because if the function doesn't return x, the x reference would disappear at the end of the function.
> 
> I was stating to question reality.

That is very unusual behaviour - there is no indication that it should be keeping the argument around. In fact, given that you are not passing an argument into the function, you would generally expect it to be the indicated empty dict.

File: idaes/core/util/scaling.py [hunk 1]
```diff
@@ -72,7 +73,7 @@ def scale_arc_constraints(blk):
             constraint_scaling_transform(c, sf)
 
 
-def map_scaling_factor(iter, default=1, warning=False, func=min):
+def map_scaling_factor(iter, default=1, warning=False, func=min, hint=None):
```
Comment (andrewlee94) [inline idaes/core/util/scaling.py:76]: Probably should document what `hint` is for - I can guess but it would be good to state it clearly.

**Review by andrewlee94 — COMMENTED** [review]
The changes to scaling look good as far as I can tell. I had a couple of minor comments on testing tolerance and other ways to do the adiabatic pressure changer.

File: idaes/generic_models/unit_models/tests/test_pressure_changer.py [hunk 2]
```diff
@@ -817,13 +816,13 @@ def test_compressor(self):
         m.fs.unit.initialize()
 
         assert m.fs.unit.costing.purchase_cost.value == \
-            pytest.approx(334598.679, abs=1e-3)
+            pytest.approx(334598.679, rel=1e-3)
```
[the remark this answers is not in the document]
Comment (eslickj) [reply idaes/generic_models/unit_models/tests/test_pressure_changer.py]: I think I only changed scaling and the answer I get is slightly off what the test says I should get.  No mater what I do to the scaling (even turning it off) I seem to get the same slightly different answer.  I'm still trying to understand this.

[the remark this answers is not in the document]
Comment (eslickj) [reply idaes/generic_models/unit_models/tests/test_pressure_changer.py]: @andrewlee94, Well it looks like I get a different answer when I run this on my own computer than what the automated tests here get, and I don't think it has anything to do with this PR.  Now I'm really confused.  I'm trying to figure out now if it's just me or if it's running a single test file vs the whole set.

[the remark this answers is not in the document]
Comment (eslickj) [reply idaes/generic_models/unit_models/tests/test_pressure_changer.py]: It took a lot of work, but it's the homotopy test.  Running the homotopy test the pressure changer or pfr test changes the result of the test.

[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/generic_models/unit_models/tests/test_pressure_changer.py]: We should fix the homotopy tests then - a test should not be making global changes to settings.

[the remark this answers is not in the document]
Comment (eslickj) [reply idaes/generic_models/unit_models/tests/test_pressure_changer.py]: Okay, it took some doing, but I think I sorted out all the issues related to test interference and solver settings here.

File: idaes/tests/test_ipopt.py [hunk 3]
```diff
@@ -36,6 +33,7 @@ def test_ipopt_idaes_config():
     """
     Test that the default solver options are set
     """
+    isolve.use_idaes_solver_configuration_deafults()
```
Comment (andrewlee94) [inline idaes/tests/test_ipopt.py:36]: Typo in method name: defaults

This reminds me that someone pointed out the "constraints" was misspelled somewhere ("constaints"?) in the scaling tools too (suffix name or storage dict?).

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
I found a couple of typos. Hopefully we can get this merged fast so we can check that it fixes #344.

**Review by andrewlee94 — DISMISSED** [review]
There is a lot to look at here, but I don't see any issues.

**Comment by eslickj** [comment]
@andrewlee94 could you take another look?  I had a minor fix, I messed up fixing a conflict with the config context PR.

**Review by andrewlee94 — DISMISSED** [review]

**Review by andrewlee94 — APPROVED** [review]

**Review by MAZamarripa — APPROVED** [review]
LGTM

**Pull request merged**